### ISSUE TITLE
CMS Revamp

### ISSUE THREAD

**Issue by chpapa** [SkygearIO/features#115]
# Description

The goal of the project is to revamp CMS, so that the new CMS fulfilled a minimal feature set of:

1. User List
2. A configurable Entity List
3. Send Push Notification with simply user segment.

# Portal Design

## Portal Features

* Have a CMS tab on the left hand-side:
   * 1st Tab: "Use the CMS" Have an intro page, with beautiful graphical introduction of features / sample to the CMS. Show the URL, and a button to open CMS in new browser. 
   * 2nd Tab: "Customisation" no idea yet.... depends on the config file... maybe a large textarea for the config file first.
   * 3rd Tab: "Access Control", a (mostly) instruction page to explain how role affect CMS user rights. And a simply textbox control to allow assign any username / email with the `cms-admin` role 
   * 4th Tab: "Whitelabel", maybe simply have a big textarea with a default CSS template?

### Wireframe
* User Management: [Skygear CMS - Users.pdf](https://github.com/SkygearIO/features/files/1375743/Skygear.CMS.-.Users.pdf)
* Push Notification: 
![skygear-notifications-old](https://user-images.githubusercontent.com/387241/31481782-be96bd70-af57-11e7-900e-7cfa62f870bb.jpg)

# API Design

## Scenario

Developer would use this feature in either one of the two ways:

1. Writing a site configuration file
2. Configuring site with code (likely javascript or the like)

### Writing a site configuration file

Developer declares their CMS site structure with a config file. The CMS plugin
should load this file and serve a site according to the content of that config file.
Config file is intended to be used in skygear-cloud portal, and also by developers
who are less confident with code / have the most common CMS requirements.

The config file is consist of the following sections:

1. `site`

   Define how the site assembles itself. Site is consist of pages. Here is a possible
   configuration (in YAML):

   ```yaml
   site:
       # Declare a built-in Dashboard page.
     - type: Dashboard
       # Declare a group of pages. It should create a folder in menu with all
       # its children layout as sub-items
     - type: Group
       label: Tables
       children:
           # Declare a page for Record "topic". Record page configuration is defined
           # in the Records section.
         - type: Record
           name: topic
         - type: Record
           name: poll
         - type: Record
           name: question
     - type: PushNotificationPanel
   ```

2. `records`

   Records define records that are mentioned in the `site` section. For every record
   type, four views (list, new, edit and destroy) can be configured.

   Each view consists of a list of fields. Field defines how Skygear data type
   shows its data in corresponding views and how they can be edited (for
   new & edit views).

   See _Built-in field types_ for the list of possible built-in fields for existing
   Skygear data types.

   See [this gist](https://gist.github.com/limouren/1edd2e164eaee2692575cd81bae2d414)
   for a YAML configuration examples.

3. `association_records`

   This section define association records that can be used in Reference and
   BackReference fields in the `records` section. Its main usage is to
   model many-to-many relationship in a relational DB.

   See _Modeling relationship_ for more details on how it is used to represent
   many-to-many relationship.

### Configuring site with code

It is expected that developers would configure CMS with code if:

1. They are more comfortable to work with code directly (i.e. the underlying layer of config file); or
2. They need to write their own custom field / page (which might requires external dependencies); or
3. They were using config file, but config file no longer fulfills their needs
   and they are forced to switch.

To make life of developers in scenario #3 easier, it is expected that a commandline
utility would help convert an existing configuration into a properly code-configured
CMS (probably not in initial phrase).

Code configuration should be a superset of config file in terms of achievable
functionalities.

_Code configuration specs examples / spec missing_

* It depends heavily on frontend stack. I haven't decided on it yet so no examples.
* I have solicited opinions from some colleagues and it seems like react
  has all the votes...

### Built-in field types

Here is a list of field types that I could think of based on existing
Skygear data types. Feel free to comment / add more of them. First level
of the list is Skygear data type:

<table>
  <thead>
    <tr>
      <th>Skygear Built-in Types</th>
      <th>CMS Field Types</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>String</td>
      <td>
        <ul>
          <li>String (single-line string)</li>
          <li>Text (multi-line string)</li>
          <li>Email</li>
          <li>URL
            <ul>
              <li>Image</li>
              <li>Video</li>
            </ul>
          </li>
          <li>(String)Dropdown</li>
          <li>Rich Text
            <ul>
              <li>HTML</li>
              <li>Markdown</li>
            </ul>
          </li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Number</td>
      <td>
        <ul>
          <li>Int</li>
          <li>Float</li>
          <li>Unix Timestamp</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Boolean</td>
      <td>Boolean</td>
    </tr>
    <tr>
      <td>JSON</td>
      <td>
        <ul>
          <li>Object</li>
          <li>Array</li>
          <li>(Object)Array (Special control about adding items?)</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Datetime</td>
      <td>
        <ul>
          <li>Datetime</li>
          <li>Date</li>
          <li>Time (Is it better to model it like a number?)</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Location</td>
      <td>
        <ul>
          <li>lat-lng Coordinate</li>
          <li>Map</li>
          <li>Address Input?</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Reference</td>
      <td>
        <ul>
          <li>Reference</li>
          <li>ReferenceArray</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Asset</td>
      <td>
        <ul>
          <li>Asset</li>
          <li>Image</li>
          <li>Video</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>ACL</td>
      <td>ACL (not in current phrase)</td>
    </tr>
  </tbody>
</table>

#### One-To-One Embedded

Consider two record types `user` and `private_user`.

```yaml
records:
  user:
    list:
      fields:
        - type: _id
        - name: name
          type: String
        - type: EmbeddedRecord
          reference:
            - name: company_id
              type: BackReference
              from: private_user
              fromField: _id  # private_user._id
              toField: _id    # user.id
              isOne: true     # denote that this back reference has only one item
          fields:
              # we need the _id field for reference purpose even if we are not
              # showing it
            - type: _id
              hidden: true
            - name: hk_id
              label: HK ID
              type: String
```

Fields in EmbeddedRecord (`private_user.hk_id`) would be shown like `user`'s fields.

#### One-To-Many

Consider a relationship between `parent` and `child`. One parent could have
many children and one child could have only one parent:

```yaml
records:
  parent:
    list:
      fields:
        - type: _id
        - name: name
          type: String
        - label: Children
          type: BackReference
          from: child
          fromField: parent_id
  child:
    list:
      fields:
        - type: _id
        - name: name
          type: String
        - name: parent_id
          type: String
        - name: parent_id
          label: Parent
          type: Reference
          referencing: parent
```

View behavior:

* list
  * `parent`: Show a link to `child` list with filter `parent_id` = current parent
  * `child`: Show `parent` id with link to its show page
* show
  * `parent`: Show a list of its `child`
  * `child`: same as list
* edit
  * `parent`: Show a table of its `child`ren. Admin could remove children. Might be
    able to add `child` with a button.
    * We probably want some form of filtering on the list of children on addition
  * `child`: Show its `parent` (might be a select). Admin should be able to switch / clear selection.
    * Again, we might want filtering on list of parents being shown.
* new
  * `parent`: A table-like interface to add existing `child` / create `child` in place.
  * `child`: Same as edit.

#### Many-To-Many

Consider two record types `country` and `enterprise`. A `country` can have
to many `enterprise`s and an `enterprise` might exist in multiple `country`.

First we define the record type used for this many-to-many relationship in
`association_records` section:

```yaml
association_records:
  country_enterprise:
    fields:
      - name: country_id
        type: Reference
        referencing: country
        referencingField: _id
      - name: enterprise_id
        type: Reference
        referencing: enterprise
        referencingField: _id
```

Now we can define this relationship in `records` section:

```yaml
records:
  country:
    list:
      fields:
        - type: _id
        - name: name
          type: String
        - name: enterprises
          type: Reference
          referencing: enterprise
          # specify association record instead of referencingField
          via_association_record: country_enterprise
  enterprise:
    list:
      fields:
        - type: _id
        - name: name
          type: String
        - name: countries
          type: Reference
          referencing: country
          via_association_record: country_enterprise

```

View behavior:

View behavior is the same as those of one-to-many _except_ that admin couldn't
create new referencing record in new page & edit page.

### Push Notification Panel

Requirements:

1. Send to all users
2. Send to users of some channels
   * Probably implemented as "Send to users satifying a query, where user.channel is an array"
   * Reuse list filter UI?
     * Developer could configure which fields can be filtered
   * Might need templating with user as context
3. Send to individual user
   * Add an built-in action s.t. admin could do this by pressing a button in list or show page?

# Open Questions

## Future Features

* Localization
  * Basically keep different versions of a field
  * Should be manually enabled for a specific field
  * Fields other than text can also be localized (e.g. Image)
  * Always has a fallback locale
  * Might be good to get locale info from user, but the basic is that caller
    has to supply it in the request (Likely HTTP Header)
  * CMS would need a way to transform Skygear response of record(s),
    probably in Go.

# Related Issues

- Server Issues
- Client Issues
- Guides Issues

# Progress Tracker

- [ ] Specification Design Approval
- [ ] Write (code + tests + API docs) then get them merged. **Put All PR there**
  - [ ] Code
  - [ ] Minimal tests
  - [ ] Minimal API Docs
- [ ] Guides. **Put All PR there**
- [ ] Release
  - [ ] Update release notes
  - [ ] Release

# Advice

- Specification Design Approval
  - Once you get LGTM from another Skygear Core Team Member, you can check this checkbox. And apply the "workflow/design-complete" label.
- Coding
  - Use as many PRs as you need.  Write tests in the same or different PRs, as is convenient for you.
  - API doc should goes in the same PR with the code.
  - As each PR is merged, add a comment to this issue referencing the PRs.
  - When you are done with the code, apply the "workflow/code-complete" label.
- Guides
  - Write or modify guides and get them merged in https://github.com/skygeario/guides
  - When the PR of guides is merged, check this checkbox and apply the "workflow/guides-complete" label.

**Comment by chpapa** [comment]
@limouren I have uploaded the old Push Notification portal design, and wireframe of user management for your reference in the Feature Description. More design to align everything is needed. And also the wireframes probably contain features that's way beyond this revamp.

**Comment by chpapa** [comment]
@limouren Looking at the initial spec, here are my thoughts:

1. actually you should check in specs and sample code in this repo :) instead of gist.

2. I guess react is good, but may also consider others component based framework so we can provide reusable component. on the other hand maybe like WordPress, if people can customize one admin page however they want, it would be even more flexible so people don't need to force to use the framework we choose / we need to becareful an update breaks everyone's page.

Regarding builtin types:
* I don't quite understand builtin type. why string have enum? I think it should be drop-down box, checkbox, auto-complete single-line, multi-line etc instead?
* what about array of objects? (which skygear builtin have it)
* time is good. yes people could have it (e.g. set a time that the person will receive daily digest)
* How reference display?
* Location wold have different type of display too... map, coordinates, address input, etc.
* ACL is tricky, maybe not in initial phase, need to think carefully

**Comment by royuen** [comment]
@limouren @chpapa 

Today I have discussed with Kenji about some future expectation, but not for this phase now. 

I believe we need to support "filter" with relationship. For example we have a "Job" listing page. It will show Job info, and also Employer info which is in another table. And for the "filter" functionality of this page (e.g. search), user should be able to choose by employer info. 

The old CMS only allows the primary table fields in filter section, which makes a lot of obstacles when doing integration.

**Comment by chpapa** [comment]
continue:

Regarding modeling relationship:

* any idea of how to add filter / query for one to many? (e.g. all jobs of a company which is still active)

* parent and child sounds confusing to me for one to many relations... (maybe even just call it one and many?)

* for one to many relations, I think we need to handle the case of some developers use "array of references". it is quite common for people not following normal form

**Comment by chpapa** [comment]
@limouren @royuen I agree that should not be in phase 1, but we should want to have it on config spec and sample first, so we make sure our config file is expandable in future

**Comment by chpapa** [comment]
@limouren @royuen I agree that should not be in phase 1, but we should want to have it on config spec and sample first, so we make sure our config file is expandable in future

**Comment by limouren** [comment]
@chpapa 

> Making PR

I intend to do that later after getting more concrete on how the config is being
mapped to the same configuration in code. I am largely outdated on current
frontend tech so I am a bit hesitated to being too detail at this early design stage.

> react or other component based framework

I have also got angular 4 and vue.js on my list.

Angular is probably the best fit as it has routing and two-way binding included.
But it requires the use of Typescript, and forcing a developer to learn a new
language in order to customize a cms doesn't sound very viable...

Vue.js is a more traditional view rendering framework and I will probably look at
it more before making decision. But React definitely has more users
than vue.js... cc @rickmak

## Built-in Type

> I don't quite understand builtin type. why string have enum? I think it should be drop-down box, checkbox, auto-complete single-line, multi-line etc instead?

Oops sorry for the ambiguity. Originally first level is skygear-server builtin
types and second level onwards are corresponding CMS field types.
I changed it to a table. Hope it is clear enough now.

I didn't go for naming field with dropdown / checkbox because they are view
oriented. There are different types that can be rendered as the same kind of view (
e.g. You can render both boolean & enum string as radio buttons) and I want to
avoid a field needing to adopt more than one kind of data type.

I didn't adopt your `display_as` as well because I think it makes extending /
creating new fields more difficult for developer. If developer wants to customize
the display of certain data type, they could just go ahead and write a Field.
IMO providing extension point in a separate attribute of Field make things a bit more
complicated than it needs to be. (e.g. we have to provide a way for developer
to register their `display_as` extension for a particular Field type)

> what about array of objects? (which skygear builtin have it)

Missed it. I added it under the JSON type.

> How reference display?

I think the "Modeling relationship" section does a fair job on how they are
displayed (not good indeed). I can write another section about it if you
need it. It should be clearly descibed in the final spec nevertheless.

> Location wold have different type of display too... map, coordinates, address input, etc.

Yes. Noted.

> ACL is tricky, maybe not in initial phase, need to think carefully

Noted too.

## Modeling Relationship

> any idea of how to add filter / query for one to many? (e.g. all jobs of a company which is still active)

For this particular case, since we are displaying the many side (jobs), it
can be solved with the default filter. I updated the gist to reflect this:
https://gist.github.com/limouren/1edd2e164eaee2692575cd81bae2d414#file-embedded-record-yaml

It is much trickier if we want to filter by foreign record's fields (e.g.
company can be soft-deleted and we only want to show those jobs where company.deleted = False).
AFAIK skygear query doesn't provide this capability. I have an idea like
developer could create a SQL view in Skygear and CMS could use that for querying,
but I haven't gone deep in it yet.

> parent and child sounds confusing to me for one to many relations... (maybe even just call it one and many?)

Parent and child is pretty standard in [doing](https://docs.jboss.org/hibernate/orm/3.5/reference/en/html/example-parentchild.html) [one-to-many](http://docs.sqlalchemy.org/en/latest/orm/basic_relationships.html) [tutorials](https://www.elastic.co/guide/en/elasticsearch/guide/current/parent-child.html).
I make it more explicit now which side is one and which side is many. Hope it is clearer now.

> for one to many relations, I think we need to handle the case of some developers use "array of references". it is quite common for people not following normal form

Ooo... I have to think about it.

**Comment by chpapa** [comment]
> I have also got angular 4 and vue.js on my list.
> 
> Angular is probably the best fit as it has routing and two-way binding included.
> But it requires the use of Typescript, and forcing a developer to learn a new
> language in order to customize a cms doesn't sound very viable...
> 
> Vue.js is a more traditional view rendering framework and I will probably look at
> it more before making decision. But React definitely has more users
> than vue.js... cc @rickmak

@limouren Please seriously consider making our architecture framework neutral for customising a new view in CMS frontend.

Giving more thought, it is actually quite a disaster if we force developers to use any JS framework. People will have love hate relations. JS frameworks tends to obsolete quickly and we need to keep maintaining it. Disaster.

Wordpress's way to allow plugins to add new view in Wordpress Admin should be better.

> Oops sorry for the ambiguity. Originally first level is skygear-server builtin
> types and second level onwards are corresponding CMS field types.
> I changed it to a table. Hope it is clear enough now.
> 
> I didn't go for naming field with dropdown / checkbox because they are view
> oriented. There are different types that can be rendered as the same kind of view (
> e.g. You can render both boolean & enum string as radio buttons) and I want to
> avoid a field needing to adopt more than one kind of data type.
> 
> I didn't adopt your display_as as well because I think it makes extending /
> creating new fields more difficult for developer. If developer wants to customize
> the display of certain data type, they could just go ahead and write a Field.
> IMO providing extension point in a separate attribute of Field make things a bit more
> complicated than it needs to be. (e.g. we have to provide a way for developer
> to register their display_as extension for a particular Field type)

@limouren We need to talk about it quickly today face to face. I think instead of enum you should just use view element as field type.

My points are:

1. It does not even make sense - So what does `enum` meant anyway in your config file? No one would know what is enum in a CMS. But Dropdown box is pretty clear.
1. CMS doesn't need `type` - When you specify a `name` for the field, Skygear CMS already know its' `type`. It is redundant to ask the developer to specify the `type` again in config file. So if there is a `type` in the config file, it's actually `display_type` (or `display_as`)
1. We will have identical display field used for different types - It doesn't matter if dropdown could be used for `string` or `boolean`. CMS needs two things for each field: display_as and type. It doesn't need to bother developer. The only thing developer need to make sure is the view element they specified in the config file is compatible with the actual DB type.

Please just change all fields types to display as, I totally can't understand the table now. And it is inconsistent. Why there are display type `Map` now but also `Enum`? The so called CMS type is `display_as` anyway. (And that's why I think we probably shouldn't name it `type` to avoid confusion too)



> Missed it. I added it under the JSON type.

Skygear doesn't have `JSON type`, it should be `Array`. Ref: https://docs.skygear.io/guides/cloud-db/data-types/js/

> I have an idea like
> developer could create a SQL view in Skygear and CMS could use that for querying,
> but I haven't gone deep in it yet.

Hmmm, so it goes back to the same idea I proposed for the config file. Maybe we should just use that? is there any reason you didn't adopt my proposed format actually?

**Comment by chpapa** [comment]
> I updated the gist to reflect this:
> https://gist.github.com/limouren/1edd2e164eaee2692575cd81bae2d414#file-embedded-record-yaml

BTW, I think this example is incorrect. My originally stated question is, sorry it wasn't specific enough in the comments, but the face-to-face discussion was exactly opposite, our use case is to display a list of companies, and click on one company will display all jobs which are active. Now this example seems displaying all jobs, and back-reference to the company.

**Comment by chpapa** [comment]
One more thing: I think now I get why I was confused by parent child, I didn't get it still because I thought it is a keywords or something. But actually how does it works? Both are first level under records. What if I want to display jobs under companies, but do not want to display jobs as a independent identities?

**Comment by limouren** [comment]
@chpapa @rickmak @royuen I tried to summerize our discussion this morning:

> On framework

1. It's okay for us to be opinionated on framework if developer wants to make
   custom field type
2. We should be framework neutral if developer want a custom page in CMS. e.g.
   We could provide a callback with a DOM node s.t. developer can append
   whatever they like.

> On field type

We will keep that for each field type it would only show one skygear data type.
Field type's name would be modified to match the display element. So there
won't be `Enum` but rather `StringDropdown`.

> JSON vs Array / Object

Will change to match the docs.

> On modeling relationship

Added the company with jobs show page [config example](https://gist.github.com/limouren/1edd2e164eaee2692575cd81bae2d414#file-embedded-record-back-ref-yaml).
